src/Compatibility/Microsoft.DotNet.ApiCompatibility/AssemblySymbolLoader.cs

```diff
@@ -0,0 +1,299 @@
+                }
+
+                if (_resolveReferences && !string.IsNullOrEmpty(directory))
+                    _referenceSearchPaths.Add(directory);
+            }
+
+            return result;
+        }
+
+        private IEnumerable<MetadataReference> LoadAssembliesFromDirectory(string directory)
+        {
+            foreach (string assembly in Directory.EnumerateFiles(directory, "*.dll"))
+            {
+                yield return CreateOrGetMetadataReferenceFromPath(assembly);
+            }
+        }
+
+        private MetadataReference CreateOrGetMetadataReferenceFromPath(string path)
+        {
+            // Roslyn doesn't support having two assemblies as references with the same identity and then getting the symbol for it.
+            string name = Path.GetFileName(path);
+            if (!_loadedAssemblies.TryGetValue(name, out MetadataReference metadataReference))
+            {
+                using FileStream stream = File.OpenRead(path);
+                metadataReference = CreateAndAddReferenceToCompilation(name, stream);
+            }
+
+            return metadataReference;
+        }
+
+        private MetadataReference CreateAndAddReferenceToCompilation(string name, Stream fileStream)
+        {
+            // If we need to resolve references we can't reuse the same stream after creating the metadata
+            // reference from it as Roslyn closes it. So instead we use PEReader and get the bytes
+            // and create the metadata reference from that.
+            using PEReader reader = new(fileStream);
+            PEMemoryBlock image = reader.GetEntireImage();
+            MetadataReference metadataReference = MetadataReference.CreateFromImage(image.GetContent());
+            _loadedAssemblies.Add(name, metadataReference);
+            _cSharpCompilation = _cSharpCompilation.AddReferences(new MetadataReference[] { metadataReference });
+
+            if (_resolveReferences)
+            {
+                ResolveReferences(reader);
+            }
+
+
+            return metadataReference;
+        }
+
+        private void ResolveReferences(PEReader peReader)
+        {
+            MetadataReader reader = peReader.GetMetadataReader();
+            foreach (AssemblyReferenceHandle handle in reader.AssemblyReferences)
+            {
+                AssemblyReference reference = reader.GetAssemblyReference(handle);
+                string name = $"{reader.GetString(reference.Name)}.dll";
+                bool found = _loadedAssemblies.TryGetValue(name, out MetadataReference _);
+                if (!found)
+                {
+                    foreach (string directory in _referenceSearchPaths)
+                    {
+                        string potentialPath = Path.Combine(directory, name);
+                        if (File.Exists(potentialPath))
+                        {
+                            // TODO: add version check and add a warning if it doesn't match?
```

Comment:
Looks similar to https://github.com/dotnet/runtime/blob/597186e83e3a1972f35c6a601fdbc889cac84b30/src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/PathAssemblyResolver.cs#L23

Public key and version are important.  If we don't constrain the load, then we need to diff them to catch things like mismatched assembly identities or lower versions in assemblies that are expected to be compatible.

Reply:
The reason why I didn't use `System.Reflection.MetadataLoadContext` to do this (that was my initial approach) is because we would need to provide the corelib path, however I think it makes sense to do it that way since users need to say "I want you to resolve references", in order for this code to be executed and if resolving references is enabled, then the search paths should contain where the corelib is?

Reply:
Will open a follow up issue.

Reply:
I didn’t mean that you had to use MLC but just consider how it solved the similar problem.  Note that we actually had a bug here and had to patch MLC because it didn’t handle retargetable assemblies.